app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts

```diff
@@ -385,9 +581,55 @@ export default class LocalBackend extends Backend {
     return Promise.resolve(null)
   }
 
-  /** Invalid operation. */
-  override createDirectory() {
-    return this.invalidOperation()
+  /** Create a directory. */
+  override async createDirectory(
+    body: backend.CreateDirectoryRequestBody
+  ): Promise<backend.CreatedDirectory> {
+    const parentDirectoryPath =
+      body.parentId == null ? this.defaultRootDirectory : extractTypeAndId(body.parentId).id
+    const path = `${parentDirectoryPath}/${body.title}`
+    await this.runProjectManagerCommand(null, 'filesystem-create-directory', path)
+    return {
+      id: newDirectoryId(path),
+      parentId: newDirectoryId(parentDirectoryPath),
+      title: body.title,
+    }
+  }
+
+  /** Change the parent directory of an asset.
+   * Changing the description is NOT supported. */
+  override async updateAsset(
+    assetId: backend.AssetId,
+    body: backend.UpdateAssetRequestBody
+  ): Promise<void> {
+    if (body.parentDirectoryId != null) {
+      const typeAndId = extractTypeAndId(assetId)
+      const sourcePath =
+        typeAndId.type === backend.AssetType.project ? body.projectPath ?? '' : typeAndId.id
+      const fileName = fileInfo.fileName(sourcePath)
+      await this.runProjectManagerCommand(
+        null,
+        'filesystem-move-from',
+        sourcePath,
+        '--filesystem-move-to',
+        `${extractTypeAndId(body.parentDirectoryId).id}/${fileName}`
+      )
+    }
+  }
+
+  /** Upload a file. */
+  override async uploadFile(
+    params: backend.UploadFileRequestParams,
+    file: Blob
+  ): Promise<backend.FileInfo> {
+    const parentPath =
+      params.parentDirectoryId == null
+        ? this.defaultRootDirectory
+        : extractTypeAndId(params.parentDirectoryId).id
+    const path = `${parentPath}/${params.fileName}`
+    await this.runProjectManagerCommand(file, 'filesystem-write-path', path)
+    // `project` MUST BE `null` as uploading projects uses a separate endpoint.
+    return { path, id: newFileId(path), project: null }
```

Comment:
If it must be `null` then why is it a field at all? Is this because `remoteBackend` provides it? Perhaps `remoteBackend` should be extending the type in that case?

Reply:
perhaps... it's possible that the current structure is partly historical baggage? in a nutshell:
- both `LocalBackend` and `RemoteBackend` are implementations of the abstract interface `Backend`. the impls aren't exposed directly, only a `Backend` of an arbitrary type
- but also the API has been designed using the `RemoteBackend` (cloud backend) as the source of truth - you can see that the `LocalBackend` does quite a bit of transformation on various endpoints to make the shape consisteent with the one returned by `RemoteBackend`